source/common/router/config_impl.cc

```diff
@@ -664,15 +672,29 @@ std::string RouteEntryImplBase::newPath(const Http::RequestHeaderMap& headers) c
     final_host = processRequestHost(headers, final_scheme, final_port);
   }
 
+  std::string final_path_value;
+  bool path_redirect_has_query = false;
   if (!path_redirect_.empty()) {
-    final_path = path_redirect_.c_str();
+    // the path_redirect query string, if any, takes precedence over the request's query string,
+    // and it will not be stripped regardless of `strip_query`.
+    absl::string_view current_path = headers.getPathValue();
+    path_redirect_has_query = path_redirect_.find('?') != absl::string_view::npos;
+    size_t path_end = current_path.find('?');
+    bool current_path_has_query = path_end != absl::string_view::npos;
+    if (!path_redirect_has_query && current_path_has_query) {
```

Comment:
If we `have path_redirect_has_query_` we can also split this if statement and skip searching headers.getPathValue() for a `?` when we know we're not going to use it.